Initial

chrisforrette opened the pull request:
#### What's this PR do?

Creates the first version of a Joi image extension, which validates image dimensions and types, and includes testing, coverage, linting, and publishing configuration.

@randallagordon One particular thing I'd like your input on is license—since this is our first open source thing and I know nothing about licenses and you do, let me know what would be best to fill in for the license in `package.json` and whether we might want to add any other indicators around that to the README or the repo.

#### Where should the reviewer start?

`src/index.js`

#### How should this be manually tested?

You can recreate the example from the README or in the unit tests if you want to try it out on your own image(s).

#### Any background context you want to provide?

For avatar upload we need to validate a minimum size so we're not stretching out avatar images, so I whipped this here Joi extension up.

#### What are the relevant tickets?

DOI-361

#### What gif best describes this PR or how it makes you feel?

https://giphy.com/gifs/mrw-dont-length-L1ZqGzJDFYC7C

#### Definition of Done:

- [x] Is there appropriate test coverage? (~80% coverage on all new code)
- [x] Is the README appropriately updated?
- [x] Are there breaking changes? Has a representative from each affected team signed-off?

chrisforrette committed:
Initial pass as Joi image extension validation

chrisforrette committed:
Installing and configuring test and coverage stuffs

chrisforrette committed:
Adding tests and ironing out test scripts

chrisforrette committed:
Adding GitHub repo info to package.json

chrisforrette committed:
Starting a README

chrisforrette committed:
Adding CircleCI config

chrisforrette committed:
Adding .npmrc token write to Circle publish job

chrisforrette committed:
Adding PR template

chrisforrette committed:
Updating coverage note in PR template

randallagordon commented on package.json:
```diff
@@ -0,0 +1,99 @@
+{
+  "name": "joi-image-extensions",
+  "version": "0.1.0",
```
Looks solid enough to jump straight to v1.0.0 to me!

chrisforrette replied:
After this gets merged I was going to just run an `npm version` to do just that.

randallagordon commented on package.json:
```diff
@@ -0,0 +1,99 @@
+    "type": "git",
+    "url": "git+https://github.com/doinginc/joi-image-extension.git"
+  },
+  "license": "UNLICENSED",
```
I'd personally go with an MIT license, though ultimately we should get @technojosh's input since its a legal question. Though it likely doesn't make sense for this package since it's associated with Joi, any future work that may produce trademarks worth protecting warrants considering a BSD 3/4-clause.

randallagordon commented on src/index.js:
```diff
@@ -0,0 +1,164 @@
+      disallowTypes: 'must be one of the following types: {{allowedTypes}}'
+    },
+
+    coerce (value, state, options) {
```
Would it be more appropriate to use `pre` here instead of `coerce`? I seem to recall there being a way/various reasons to turn off coercion and I think `pre` would still run regardless. Mainly thinking for chaining purposes where something else may need to turn off coercion for some reason, say an input form or CLI that specifies dimensions.

chrisforrette replied:
@randallagordon Indeed, it seems that way—good call.

randallagordon commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,29 @@
+
+- [ ] Is there appropriate test coverage? (~80% coverage)
+- [ ] Is the README appropriately updated?
+- [ ] Are there breaking changes? Has a representative from each affected team signed-off?
```
The wording here is a bit odd for an open source project, maybe drop the second question in favor of a statement such as "Please add notes about what they are and why the change is necessary."?

randallagordon commented on README.md:
```diff
@@ -1,2 +1,89 @@
+
+## API
+
+### `image.minWidth(minimum pixel width)`
```
When looking at the README as rendered by GitHub rather than as Markdown, this formatting looks a bit odd as the backtick usage implies a function signature. Maybe camelcase or just shorten them since the description immediately after explains it, `image.minWidth(minimumPixelWidth)`/`image.minWidth(minimum)`?

chrisforrette replied:
@randallagordon I agree, it's kinda weird, but I'm just following suit with how Joi documents its stuff, which seems to be `(namespace).(function)`, e.g.: https://github.com/hapijs/joi/blob/v13.1.0/API.md#anyallowvalue

randallagordon replied:
They don't break them up with spaces such that it is displaying invalid code, however:

![image](https://user-images.githubusercontent.com/635388/34749410-21086500-f556-11e7-9cdb-f139fe15f48f.png)

vs:

![image](https://user-images.githubusercontent.com/635388/34749404-167a6f0c-f556-11e7-8b98-924e39a5ee42.png)

randallagordon reviewed (changes requested):
This is lookin' rad! 💥  Just a handful of house-clean-y kinds of things.

chrisforrette committed:
Adding MIT license file and setting it in package.json, removing duplicate license key from packge.json, changing PR template verbiage

chrisforrette committed:
Moving test requirement from job into workflow for CircleCI, fixing some publish command steps, making semver tag pattern more strict

chrisforrette commented:
@randallagordon OK this should be all set—I've addressed all of your comments and I went ahead and just added an MIT license. The CircleCI stuff might need to be jiggered with once I get a tag flowing into it, and I think this PR is not running on it because I didn't enable it until after opening this.

chrisforrette committed:
Correcting package name, using 'pre' instead of 'coerce' for caching image size
